Add 8 interfaces

BoBeauf opened the pull request:
Add 8 preferences interfaces.

BoBeauf committed:
Add 8 interfaces

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/Preference.java:
```diff
@@ -0,0 +1,42 @@
+ * Preference Interface is used to define the order chosen by a Voter between
+ * some alternatives.
+ *
+ * default voter : voter 0
```
Default voter est à préciser sur `getVoter`: Returns the voter 0 if no specific voter is associated to this preference.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/Preference.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * @return number of considered alternatives
+     */
+    public int size();
```
Ceci a été retiré du diagramme (ou aurait dû l’être suite à mes derniers commentaires).

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/AntiSymmetricPreference.java:
```diff
@@ -0,0 +1,12 @@
+public interface AntiSymmetricPreference extends Preference {
+
+    /*
+    * Add here more features if needed
```
Surcharger `#asGraph()` pour en changer le contrat : préciser que le graphe renvoyé est garanti antisymétrique.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/MutablePreference.java:
```diff
@@ -0,0 +1,19 @@
+
+/**
+ * A mutable preference is a preference in which some alternatives can be added.
+ * However, It's not possible to modify the alternatives' order.
```
… in which some edges (pair of alternatives) can be added. Retirer la deuxième phrase, pas très claire. Vous pouvez éventuellement préciser : if an alternative a is strictly better than an alternative b, it is impossible to change this afterwards.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/MutableAntiSymmetricPreference.java:
```diff
@@ -0,0 +1,26 @@
+
+/**
+ * a mutable anti-symmetric preference is an anti-symmetric preference (without
+ * equal alternatives) in which some alternatives can be added.
```
An alternative equals another one iff they have the same id. What you mean is: without alternatives considered “ex-æquo”.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/MutableAntiSymmetricPreference.java:
```diff
@@ -0,0 +1,26 @@
+    public void addAlternative(Alternative a);
+
+    /**
+     * addStrictPreference adds to the preference "a1>a2";
```
Don’t repeat the name of the method

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/Preference.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * @return the Graph corresponding to the Preference
+     */
+    public Graph<Alternative> asGraph();
```
Préciser : the returned graph reads through this object: if this object is mutable, any modification to this object modifies the returned graph, and conversely. Même chose pour `getAlternatives`.
Préciser également : the graph represents the “at least as good as” relation. It is reflexive.

oliviercailloux replied:
Retirer tous ces br, inverser pour expliquer d’abord.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/CompletePreference.java:
```diff
@@ -0,0 +1,41 @@
+
+    /**
+     *
+     * @return same data but in an Immutable list object
```
Il est important de préciser ici l’ordre des rangs ! La première entrée de la liste (numéro zéro) est la meilleure classe d’équivalence : les alternatives qui n’ont aucune alternatives strictement préférées à elles.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/CompletePreference.java:
```diff
@@ -0,0 +1,41 @@
+    /**
+     *
+     * @param a is an alternative
+     * @return the rank of this alternative.
```
Préciser : a number between 1 and n, where n is the number of alternatives. Mettre n en italique. (Lettre habituellement utilisée pour représenter le nombre d’alternatives.)
@throws IllegalArgumentException if <code>a</code> is not contained in this preference.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/CompletePreference.java:
```diff
@@ -0,0 +1,41 @@
+     * @param n is a rank
+     * @return the alternative at this rank
+     */
+    public Alternative getAlternative(int n);
```
ImmutableSet<Alternative> getAlternatives(int rank), comme déjà suggéré
Préciser comme au-dessus que le rang commence à un, préciser qu’un IllegalArgumentException est lancé si le rang est ≤ 0, et que s’il n’y a pas d’alternative à ce rang, un ensemble vide est renvoyé.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/LinearPreference.java:
```diff
@@ -0,0 +1,20 @@
+
+    /**
+     * 
+     * @return a sorted list of alternatives corresponding to the preference
```
, with the best alternatives coming first.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/Preference.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * Preference Interface is used to define the order chosen by a Voter between
+ * some alternatives.
+ *
```
Préciser : a preference is empty iff it contains no alternative, equivalently, iff it contains no edge.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/LinearPreference.java:
```diff
@@ -0,0 +1,20 @@
+import io.github.oliviercailloux.j_voting.Alternative;
+
+/**
+ * A linear preference is an anti-symmetric complete preference. A linear
```
Écrire systématiquement antisymmetric (vérifier partout).

oliviercailloux reviewed (changes requested):
En lien avec PR #17.

mloubet committed:
+ Corrected all the details according to all the comments on last PR.
+--> Corrected javadoc
+--> Corredted some types
+--> deleted one method

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/CompletePreference.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @param a is an <code>Alternative</code>
+     * @return the rank of this alternative (a number between 1 and <i>n</i>
+     * where <i>n</i> is the total number of <code>Alternative</code><br><br>
```
Enlever les `<br>`, la mise en page est le boulot de l’outil de génération de doc. Fermer la parenthèse. Écrire « \<code>Alternative\</code> instances » ou plus simplement « alternatives ».

Il faut aussi préciser comment les rangs sont comptés car il y a ambiguïté en cas d’ex-æquo. Écrire par exemple : the rank of an alternative is one plus the number of alternatives strictly preferred to it.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/CompletePreference.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     *
+     * @return Same data but in an Immutable list object<br>
+     * Order in the list matters : it means that a set of alternative is
```
Retirer « Order in the list matters » (c’est sous-entendu par le choix d’une List comme type de retour). Retirer « it means that ».

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/ImmutableAntiSymmetricPreference.java:
```diff
@@ -0,0 +1,14 @@
+*
+* Add here more features if needed
+*/
+	
```
Il faut donc surcharger `asGraph()` pour préciser son contrat.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/ImmutablePreference.java:
```diff
@@ -0,0 +1,24 @@
+    /**
+     * @return the Graph corresponding to the Preference
+     */
+    @Override
```
Ne pas remplacer la doc par une doc moins précise ! La doc de `Preference#asGraph()` convient très bien. (Ne pas la copier pour autant !)

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/ImmutablePreference.java:
```diff
@@ -0,0 +1,24 @@
+    /**
+     * @return alternatives's set
+     */
+    @Override
```
DRY

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/LinearPreference.java:
```diff
@@ -0,0 +1,22 @@
+    /**
+     * 
+     * @return a sorted list of alternatives corresponding to the preference.
+     * <br>
```
Retirer `<br>`. Rédigez dans le premier § si vous voulez. Inspirez-vous de la javadoc de `List` ou `ImmutableList` par exemple.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/MutableAntiSymmetricPreference.java:
```diff
@@ -0,0 +1,26 @@
+
+/**
+ * a mutable antisymmetric preference is an antisymmetric preference (without
+ * alternatives considered ex-aequo) in which some alternatives can be added.
```
in which alternatives and edges can be added.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/MutablePreference.java:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * A mutable preference is a preference in which some alternatives can be added,
+ * which means some edges (pair of alternatives) can be added to the
```
in which edges (pairs of alternatives) and alternatives can be added. (Retirer to the corresponding graph). Retirer br, remplacer par paragraphes

oliviercailloux replied:
Retirer to the corresponding graph

FORMAT (et les § doivent rester !)

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/MutablePreference.java:
```diff
@@ -0,0 +1,22 @@
+
+    /**
+     * 
+     * @return a MutableGraph of alternatives who represents the preference.
```
Trop imprécis. Le graphe est-il réflexif ? Etc. Explorer usage de inheritDoc.

mloubet committed:
+ Updated every Preference interface to comply with all the comments on last PR.
+--> Used {inheritDoc}
+--> deleted all <br>
+--> corrected some javadocs to make it more accurate

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/ImmutableAntiSymmetricPreference.java:
```diff
@@ -0,0 +1,21 @@
+/* This interface combines the ImutablePreference and AntiSymetricPreference
+*
+* Add here more features if needed
+*/
```
Retirer ce commentaire inutile.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/ImmutablePreference.java:
```diff
@@ -0,0 +1,31 @@
+    /**
+     *
+     * In the graph : a relation from <i>a</i> to <i>b</i> means "a is at least
+     * as good as b".
```
Ceci est valable pour toute préférence. Déplacer le commentaire et utiliser inheritDoc.

oliviercailloux commented on src/main/java/io/github/oliviercailloux/j_voting/preferences/interfaces/LinearPreference.java:
```diff
@@ -0,0 +1,22 @@
+     * 
+     * @return a sorted list of alternatives corresponding to the preference.
+     *
+     * Best alternatives coming first in the list.
```
Inverser pour respecter la convention (d’abord explications, ensuite @return).

mloubet committed:
+ Updated after your comments

oliviercailloux reviewed (approved):
J’imagine que c’est bon maintenant.
